application/controllers/template_controller.py

```diff
@@ -0,0 +1,64 @@
+from application.models.models import CommunicationTemplate
+from application.utils.session_wrapper import with_db_session
+from application.utils.exceptions import InvalidTemplateException, DatabaseError
+from application.models.schema import template_schema
+from jsonschema import validate, ValidationError
+from structlog import get_logger
+
+
+logger = get_logger()
+
+UPLOAD_SUCCESSFUL = 'The upload was successful'
+INVALID_TEMPLATE = 'Template is invalid'
+PREEXISTING_TEMPLATE = 'Id already exists'
+
+
+def get_template_by_id(template_id, session):
+    return session.query(CommunicationTemplate).filter(CommunicationTemplate.id == template_id).first()
+
+
+def validate_template(template):
+    try:
+        validate(template, template_schema)
+    except ValidationError as exception:
+        logger.exception("Attempted to upload invalid template")
+        raise InvalidTemplateException(INVALID_TEMPLATE, status_code=400)
+
+
+class TemplateController(object):
+
+    @staticmethod
+    @with_db_session
+    def upload_comms_template(template_id, template_object, session=None):
+        logger.info('Uploading comms template with id {}.'.format(template_id))
+
+        validate_template(template_object)
+
+        existing_template = get_template_by_id(template_id, session)
+
+        if existing_template:
+            logger.info("Attempted to upload already existing template, id {}".format(template_id))
+            raise InvalidTemplateException(PREEXISTING_TEMPLATE, status_code=400)
+
+        label = template_object.get('label')
+        type = template_object.get('type')
+        uri = template_object.get('uri')
+        classification = template_object.get('classification')
+        params = template_object.get('params')
+
+        template = CommunicationTemplate(id=template_id, label=label, type=type, uri=uri, classification=classification,
+                                         params=params)
+
+        session.add(template)
+
+        logger.info("Uploaded template with id {}".format(template_id))
+
+        return UPLOAD_SUCCESSFUL
+
+    @staticmethod
+    @with_db_session
+    def get_comms_template_by_id(template_id, session=None):
+        template = get_template_by_id(template_id, session)
+        if not template:
+            raise DatabaseError("Template with id {} doesn't exist".format(template_id), status_code=404)
```

Comment:
Should this really be an exception? Could we not return None and handle this in the view?

Reply:
I wasn't sure about this as I didn't want logic in the view.

Reply:
It seems like view level logic to decide what http status to return though is all that I'd say